From 0b786419365b9c3c297306884dce30ad2d890fb6 Mon Sep 17 00:00:00 2001
From: Dhairya Khale <dkhale@qti.qualcomm.com>
Date: Sun, 12 Jul 2026 19:19:47 +0530
Subject: [PATCH] net: Assign custom buffer from netdev_alloc_skb_fast and
 remove recycler alloc from netdev_alloc_skb

Change-Id: I9ab32fce911587ed524846125cf8e5a7e9e49337
Signed-off-by: Dhairya Khale <dkhale@qti.qualcomm.com>
(cherry picked from commit c811803eebad4a2f36555448ada8dc23af55818e)
---
 include/linux/gfp_types.h      | 13 ++++++++++++
 include/linux/skbuff.h         |  4 +++-
 include/trace/events/mmflags.h |  8 ++++++++
 net/core/skbuff.c              | 37 +++++++++-------------------------
 4 files changed, 33 insertions(+), 29 deletions(-)

diff --git a/include/linux/gfp_types.h b/include/linux/gfp_types.h
index dfde1e1e321c..c3aac2e3a324 100644
--- a/include/linux/gfp_types.h
+++ b/include/linux/gfp_types.h
@@ -60,6 +60,11 @@ typedef unsigned int __bitwise gfp_t;
 #else
 #define ___GFP_NOLOCKDEP	0
 #endif
+#ifdef CONFIG_SKB_RECYCLER
+#define ___GFP_RECYCLER		0x8000000u
+#else
+#define ___GFP_RECYCLER		0
+#endif
 /* If the above are modified, __GFP_BITS_SHIFT may need updating */
 
 /*
@@ -250,8 +255,16 @@ typedef unsigned int __bitwise gfp_t;
 /* Disable lockdep for GFP context tracking */
 #define __GFP_NOLOCKDEP ((__force gfp_t)___GFP_NOLOCKDEP)
 
+/* Enabling recycler flag */
+#define __GFP_RECYCLER ((__force gfp_t)___GFP_RECYCLER)
+
 /* Room for N __GFP_FOO bits */
+#ifdef CONFIG_SKB_RECYCLER
+#define __GFP_BITS_SHIFT 28
+#else
 #define __GFP_BITS_SHIFT (26 + IS_ENABLED(CONFIG_LOCKDEP))
+#endif
+
 #define __GFP_BITS_MASK ((__force gfp_t)((1 << __GFP_BITS_SHIFT) - 1))
 
 /**
diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 7dc1fbe5b387..35df8d005d1a 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -1065,6 +1065,8 @@ struct sk_buff {
 	/* Packets processed in dev_fast_xmit_qdisc() path */
 	__u8			int_pri:4;
 	/* Priority info for hardware qdiscs */
+	__u8			is_from_custom_cache:1;
+	/* Allocated from custom cache with recycler enabled */
 
 #ifdef CONFIG_IPQ_PON
 	/* packets processed by pon_iftrap_......() */
@@ -3416,7 +3418,7 @@ static inline struct sk_buff *netdev_alloc_skb(struct net_device *dev,
 static inline struct sk_buff *netdev_alloc_skb_fast(struct net_device *dev,
 						    unsigned int length)
 {
-	return __netdev_alloc_skb_fast(dev, length, GFP_ATOMIC);
+	return __netdev_alloc_skb_fast(dev, length, GFP_ATOMIC | __GFP_RECYCLER);
 }
 
 /* legacy helper around __netdev_alloc_skb() */
diff --git a/include/trace/events/mmflags.h b/include/trace/events/mmflags.h
index e010618f9326..d3389988c491 100644
--- a/include/trace/events/mmflags.h
+++ b/include/trace/events/mmflags.h
@@ -60,9 +60,17 @@
 #define __def_gfpflag_names_kasan
 #endif
 
+#ifdef CONFIG_SKB_RECYCLER
+#define __def_gfpflag_names_recycler ,			\
+	gfpflag_string(__GFP_RECYCLER)
+#else
+#define __def_gfpflag_names_recycler
+#endif
+
 #define show_gfp_flags(flags)						\
 	(flags) ? __print_flags(flags, "|",				\
 	__def_gfpflag_names __def_gfpflag_names_kasan			\
+	__def_gfpflag_names_recycler					\
 	) : "none"
 
 #ifdef CONFIG_MMU
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 8e17dbfc098e..37163fbce873 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -730,7 +730,12 @@ static void *kmalloc_reserve(unsigned int *size, gfp_t flags, int node,
 	unsigned int cache_size;
 
 	obj_size = SKB_HEAD_ALIGN(*size);
-	if ((obj_size <= SKB_SMALL_HEAD_CACHE_SIZE &&
+	if (flags & __GFP_RECYCLER) {
+		obj = kmem_cache_alloc_node(skb_data_cache, flags | __GFP_NOMEMALLOC |	\
+					__GFP_NOWARN, node);
+		*size = SKB_DATA_CACHE_SIZE;
+		goto out;
+	} else if ((obj_size <= SKB_SMALL_HEAD_CACHE_SIZE &&
 	    !(flags & KMALLOC_NOT_NORMAL_BITS)) ||
 	    (obj_size > SZ_2K && obj_size <= SKB_DATA_CACHE_SIZE_2100)) {
 		cache = skb_select_data_cache(obj_size, &cache_size);
@@ -879,31 +884,6 @@ struct sk_buff *__netdev_alloc_skb(struct net_device *dev,
 {
 	struct sk_buff *skb;
 	unsigned int len = length;
-
-#ifdef CONFIG_SKB_RECYCLER
-	bool reset_skb = true;
-	skb = skb_recycler_alloc(dev, length, reset_skb);
-	if (likely(skb)) {
-		skb_recycler_clear_flags(skb);
-#ifdef CONFIG_DEBUG_KMEMLEAK
-		kmemleak_update_trace(skb);
-		kmemleak_restore(skb, 1);
-		kmemleak_update_trace(skb->head);
-		kmemleak_restore(skb->head, 1);
-#endif
-		return skb;
-	}
-
-	len = SKB_RECYCLE_SIZE;
-	if (unlikely(length > SKB_RECYCLE_SIZE))
-		len = length;
-
-	skb = __alloc_skb(len + NET_SKB_PAD, gfp_mask,
-			  SKB_ALLOC_RX, NUMA_NO_NODE);
-	if (!skb)
-		goto skb_fail;
-	goto skb_success;
-#else
 	struct page_frag_cache *nc;
 	bool pfmemalloc;
 	bool page_frag_alloc_enable = true;
@@ -911,7 +891,6 @@ struct sk_buff *__netdev_alloc_skb(struct net_device *dev,
 
 	len += NET_SKB_PAD;
 
-
 #ifdef CONFIG_ALLOC_SKB_PAGE_FRAG_DISABLE
 	page_frag_alloc_enable = false;
 #endif
@@ -957,7 +936,6 @@ struct sk_buff *__netdev_alloc_skb(struct net_device *dev,
 	if (pfmemalloc)
 		skb->pfmemalloc = 1;
 	skb->head_frag = 1;
-#endif
 
 skb_success:
 	skb_reserve(skb, NET_SKB_PAD);
@@ -1004,6 +982,7 @@ struct sk_buff *__netdev_alloc_skb_fast(struct net_device *dev,
 	if (!skb)
 		goto skb_fail;
 
+	skb->is_from_custom_cache = 1;
 	goto skb_success;
 #else
 	struct page_frag_cache *nc;
@@ -1114,11 +1093,13 @@ struct sk_buff *__netdev_alloc_skb_no_skb_reset(struct net_device *dev,
 	if (unlikely(length > SKB_RECYCLE_SIZE))
 		len = length;
 
+	gfp_mask |= __GFP_RECYCLER;
 	skb = __alloc_skb(len + NET_SKB_PAD, gfp_mask,
 				SKB_ALLOC_RX, NUMA_NO_NODE);
 	if (!skb)
 		return NULL;
 
+	skb->is_from_custom_cache = 1;
 	skb_reserve(skb, NET_SKB_PAD);
 	skb->dev = dev;
 	return skb;
-- 
2.34.1

